batch/utils/copy_files.py

```diff
@@ -0,0 +1,342 @@
+        total = finish_time - self.start_time
+        if exc_type is None:
+            print(f'copied {self.src} to {self.dest} in {total:.3f}s')
+        else:
+            print(f'failed to copy {self.src} to {self.dest} in {total:.3f}s due to {exc!r}')
+
+
+class FilePart(io.IOBase):
+    def __init__(self, filename, offset, length):
+        self._fp = open(filename, 'rb')
+        self.length = length
+        self._start = offset
+        self._end = self._start + self.length
+        self._fp.seek(self._start)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+    def tell(self):
+        return self._fp.tell() - self._start
+
+    def read(self, size=-1):
+        if size < 0:
+            size = self.length
+        size = min(size, self._end - self._fp.tell())
+        return self._fp.read(max(0, size))
+
+    def seek(self, offset, whence=os.SEEK_SET):
+        if whence == os.SEEK_END:
+            return self._fp.seek(offset + self._end)
+        elif whence == os.SEEK_CUR:
+            return self._fp.seek(offset, whence)
+        else:
+            assert whence == os.SEEK_SET
+            return self._fp.seek(self._start + offset)
+
+    def close(self):
+        self._fp.close()
+
+    def flush(self, size=None):
+        raise NotImplementedError('flush is not implemented in FilePart.')
+
+    def fileno(self, size=None):
+        raise NotImplementedError('fileno is not implemented in FilePart.')
+
+    def isatty(self, size=None):
+        raise NotImplementedError('isatty is not implemented in FilePart.')
+
+    def next(self, size=None):
+        raise NotImplementedError('next is not implemented in FilePart.')
+
+    def readline(self, size=None):
+        raise NotImplementedError('readline is not implemented in FilePart.')
+
+    def readlines(self, size=None):
+        raise NotImplementedError('readlines is not implemented in FilePart.')
+
+    def xreadlines(self, size=None):
+        raise NotImplementedError('xreadlines is not implemented in FilePart.')
+
+    def truncate(self, size=None):
+        raise NotImplementedError('truncate is not implemented in FilePart.')
+
+    def write(self, size=None):
+        raise NotImplementedError('write is not implemented in FilePart.')
+
+    def writelines(self, size=None):
+        raise NotImplementedError('writelines is not implemented in FilePart.')
+
+
+def is_gcs_path(file):
+    return file.startswith('gs://')
+
+
+def flatten(its):
+    return [x for it in its for x in it]
+
+
+def listdir(path):
+    if not os.path.exists(path):
+        raise FileNotFoundError(path)
+    if os.path.isfile(path):
+        return [(path, os.path.getsize(path))]
+    # gsutil doesn't copy empty directories
+    return flatten([listdir(path.rstrip('/') + '/' + f) for f in os.listdir(path)])
+
+
+def get_dest_path(file, src, include_recurse_dir):
+    src = src.rstrip('/').split('/')
```

Comment:
I don't think this function is a good idea and makes things more complicated. GCS paths can have double slashes and they are distinct from single slash paths whereas unix consolidates them. I just wanted to strip the trailing slash when comparing whether the path was the same length. 

Example:
file = '/a/b/c/d.txt'
src = '/a/b/c/'

If you don't strip the trailing slash on the src, file and src will be the same length and you'll copy it to the wrong place (d.txt instead of c/d.txt).